File: local-server/routes/quick_search.js
/**
 * Quick Search - Search across all collections
 * Desktop + Local Server route
 */
const express = require('express');

module.exports = function quickSearchRoutes(database) {
  const router = express.Router();

  router.get('/api/quick-search', (req, res) => {
    const q = (req.query.q || '').trim();
    const limit = Math.min(parseInt(req.query.limit) || 5, 20);

    if (!q) return res.json({ results: [], total: 0 });

    const results = [];
    const ql = q.toLowerCase();
    const match = (val) => val && String(val).toLowerCase().includes(ql);

    // 1. Mill Entries
    (database.data.entries || [])
      .filter(e => match(e.truck_no) || match(e.agent_name) || match(e.mandi_name) || match(e.rst_no) || match(e.tp_no))
      .sort((a, b) => new Date(b.created_at || 0) - new Date(a.created_at || 0))
      .slice(0, limit)
      .forEach(e => results.push({
        id: e.id, type: 'entry', tab: 'entries',
        title: `Truck: ${e.truck_no || ''}`,
        subtitle: `${e.agent_name || ''} - ${e.mandi_name || ''} | ${e.kg || 0} kg`,
        date: e.date || '', data: { id: e.id, date: e.date, truck_no: e.truck_no, agent_name: e.agent_name, mandi_name: e.mandi_name, kg: e.kg, qntl: e.qntl }
      }));

    // 2. Cash Transactions
    (database.data.cash_transactions || [])
      .filter(t => match(t.category) || match(t.description) || match(t.party_type))
      .sort((a, b) => new Date(b.created_at || 0) - new Date(a.created_at || 0))
      .slice(0, limit)
      .forEach(t => results.push({
        id: t.id, type: 'cash_transaction', tab: 'cashbook',
        title: `${t.category || ''} (${t.txn_type === 'jama' ? 'Jama' : 'Nikasi'})`,
        subtitle: `Rs.${(t.amount || 0).toLocaleString('en-IN')} | ${(t.description || '').slice(0, 60)}`,
        date: t.date || '', data: { id: t.id, date: t.date, category: t.category, amount: t.amount, txn_type: t.txn_type, account: t.account }
      }));

    // 3. Private Paddy
    (database.data.private_paddy || [])
      .filter(p => match(p.party_name) || match(p.mandi_name))
      .sort((a, b) => new Date(b.created_at || 0) - new Date(a.created_at || 0))
      .slice(0, limit)
      .forEach(p => results.push({
        id: p.id, type: 'private_paddy', tab: 'payments',
        title: `Private: ${p.party_name || ''}`,
        subtitle: `${p.mandi_name || ''} | Total: Rs.${(p.total_amount || 0).toLocaleString('en-IN')}`,
        date: p.date || '', data: { id: p.id, date: p.date, party_name: p.party_name, mandi_name: p.mandi_name, total_amount: p.total_amount }
      }));

    // 4. Sale Vouchers
    (database.data.sale_vouchers || [])
      .filter(s => match(s.party_name) || match(s.voucher_no) || match(s.truck_no))
      .sort((a, b) => new Date(b.created_at || 0) - new Date(a.created_at || 0))
      .slice(0, limit)
      .forEach(s => results.push({
        id: s.id, type: 'sale_voucher', tab: 'vouchers',
        title: `Sale: ${s.party_name || ''}`,
        subtitle: `Voucher #${s.voucher_no || ''} | Rs.${(s.total || 0).toLocaleString('en-IN')}`,
        date: s.date || '', data: { id: s.id, date: s.date, party_name: s.party_name, voucher_no: s.voucher_no, total: s.total }
      }));

    // 5. Purchase Vouchers
    (database.data.purchase_vouchers || [])
      .filter(p => match(p.party_name) || match(p.voucher_no) || match(p.truck_no))
      .sort((a, b) => new Date(b.created_at || 0) - new Date(a.created_at || 0))
      .slice(0, limit)
      .forEach(p => results.push({
        id: p.id, type: 'purchase_voucher', tab: 'vouchers',
        title: `Purchase: ${p.party_name || ''}`,
        subtitle: `Voucher #${p.voucher_no || ''} | Rs.${(p.total || 0).toLocaleString('en-IN')}`,
        date: p.date || '', data: { id: p.id, date: p.date, party_name: p.party_name, voucher_no: p.voucher_no, total: p.total }
      }));

    // 6. DC Entries
    (database.data.dc_entries || [])
      .filter(d => match(d.party_name) || match(d.vehicle_no) || match(d.lot_no))
      .sort((a, b) => new Date(b.created_at || 0) - new Date(a.created_at || 0))
      .slice(0, limit)
      .forEach(d => results.push({
        id: d.id, type: 'dc_entry', tab: 'dctracker',
        title: `DC: ${d.party_name || ''}`,
        subtitle: `Vehicle: ${d.vehicle_no || ''} | Lot: ${d.lot_no || ''}`,
        date: d.date || '', data: { id: d.id, date: d.date, party_name: d.party_name, vehicle_no: d.vehicle_no, lot_no: d.lot_no }
      }));

    // 7. Staff
    (database.data.staff || [])
      .filter(s => match(s.name) || match(s.phone) || match(s.role))
      .slice(0, limit)
      .forEach(s => results.push({
        id: s.id, type: 'staff', tab: 'staff',
        title: `Staff: ${s.name || ''}`,
        subtitle: `${s.role || ''} | Rs.${(s.salary || 0).toLocaleString('en-IN')}`,
        date: '', data: { id: s.id, name: s.name, role: s.role, salary: s.salary }
      }));

    // 8. Milling Entries
    (database.data.milling_entries || [])
      .filter(m => match(m.rice_type) || match(m.note))
      .sort((a, b) => new Date(b.created_at || 0) - new Date(a.created_at || 0))
      .slice(0, limit)
      .forEach(m => results.push({
        id: m.id, type: 'milling', tab: 'milling',
        title: `Milling: ${m.rice_type || ''}`,
        subtitle: `Paddy: ${m.paddy_input_qntl || 0}Q | Rice: ${m.rice_qntl || 0}Q | OT: ${m.outturn_ratio || 0}%`,
        date: m.date || '', data: { id: m.id, date: m.date, rice_type: m.rice_type, paddy_input_qntl: m.paddy_input_qntl, rice_qntl: m.rice_qntl }
      }));

    // 9. Diesel Accounts
    (database.data.diesel_accounts || [])
      .filter(d => match(d.truck_no) || match(d.pump_name) || match(d.description))
      .sort((a, b) => new Date(b.created_at || 0) - new Date(a.created_at || 0))
      .slice(0, limit)
      .forEach(d => results.push({
        id: d.id, type: 'diesel', tab: 'cashbook',
        title: `Diesel: ${d.truck_no || ''}`,
        subtitle: `${d.pump_name || ''} | Rs.${(d.amount || 0).toLocaleString('en-IN')}`,
        date: d.date || '', data: { id: d.id, date: d.date, truck_no: d.truck_no, pump_name: d.pump_name, amount: d.amount }
      }));

    // 10. Mill Parts Stock
    (database.data.mill_parts_stock || [])
      .filter(m => match(m.part_name) || match(m.party_name))
      .sort((a, b) => new Date(b.created_at || 0) - new Date(a.created_at || 0))
      .slice(0, limit)
      .forEach(m => results.push({
        id: m.id, type: 'mill_part', tab: 'mill-parts',
        title: `Part: ${m.part_name || ''}`,
        subtitle: `${m.party_name || ''} | Rs.${(m.total_amount || 0).toLocaleString('en-IN')}`,
        date: m.date || '', data: { id: m.id, date: m.date, part_name: m.part_name, party_name: m.party_name, total_amount: m.total_amount }
      }));

    // 11. Hemali Payments
    (database.data.hemali_payments || [])
      .filter(h => match(h.sardar_name))
      .sort((a, b) => new Date(b.created_at || 0) - new Date(a.created_at || 0))
      .slice(0, limit)
      .forEach(h => results.push({
        id: h.id, type: 'hemali', tab: 'hemali',
        title: `Hemali: ${h.sardar_name || ''}`,
        subtitle: `Rs.${(h.total || 0).toLocaleString('en-IN')} | ${h.status || ''}`,
        date: h.date || '', data: { id: h.id, date: h.date, sardar_name: h.sardar_name, total: h.total, status: h.status }
      }));

    // 12. Rice Sales
    (database.data.rice_sales || [])
      .filter(r => match(r.buyer_name) || match(r.party_name))
      .sort((a, b) => new Date(b.created_at || 0) - new Date(a.created_at || 0))
      .slice(0, limit)
      .forEach(r => results.push({
        id: r.id, type: 'rice_sale', tab: 'payments',
        title: `Rice Sale: ${r.buyer_name || r.party_name || ''}`,
        subtitle: `Rs.${(r.total_amount || 0).toLocaleString('en-IN')}`,
        date: r.date || '', data: { id: r.id, date: r.date, buyer_name: r.buyer_name, total_amount: r.total_amount }
      }));

    // 13. Truck Leases
    (database.data.truck_leases || [])
      .filter(t => match(t.truck_no) || match(t.owner_name))
      .slice(0, limit)
      .forEach(t => results.push({
        id: t.id, type: 'truck_lease', tab: 'payments',
        title: `Lease: ${t.truck_no || ''}`,
        subtitle: `Owner: ${t.owner_name || ''} | Rent: Rs.${(t.monthly_rent || 0).toLocaleString('en-IN')}`,
        date: '', data: { id: t.id, truck_no: t.truck_no, owner_name: t.owner_name, monthly_rent: t.monthly_rent }
      }));

    res.json({ results, total: results.length, query: q });
  });

  return router;
};

File: backend/routes/quick_search.py
from fastapi import APIRouter, Query
from database import db
from typing import Optional
import re

router = APIRouter()


@router.get("/quick-search")
async def quick_search(q: str = Query(..., min_length=1), limit: int = Query(5, ge=1, le=20)):
    """Search across all collections. Returns grouped results."""
    if not q or len(q.strip()) < 1:
        return {"results": [], "total": 0}

    q = q.strip()
    rgx = {"$regex": re.escape(q), "$options": "i"}
    results = []

    # 1. Mill Entries
    entries_cur = db["mill_entries"].find(
        {"$or": [
            {"truck_no": rgx}, {"agent_name": rgx}, {"mandi_name": rgx},
            {"rst_no": rgx}, {"tp_no": rgx}
        ]},
        {"_id": 0, "id": 1, "date": 1, "truck_no": 1, "agent_name": 1, "mandi_name": 1, "rst_no": 1, "kg": 1, "qntl": 1}
    ).sort("created_at", -1).limit(limit)
    async for e in entries_cur:
        results.append({
            "id": e.get("id"), "type": "entry", "tab": "entries",
            "title": f"Truck: {e.get('truck_no', '')}",
            "subtitle": f"{e.get('agent_name', '')} - {e.get('mandi_name', '')} | {e.get('kg', 0)} kg",
            "date": e.get("date", ""), "data": e
        })

    # 2. Cash Transactions
    cash_cur = db["cash_transactions"].find(
        {"$or": [{"category": rgx}, {"description": rgx}, {"party_type": rgx}]},
        {"_id": 0, "id": 1, "date": 1, "category": 1, "description": 1, "amount": 1, "txn_type": 1, "account": 1}
    ).sort("created_at", -1).limit(limit)
    async for t in cash_cur:
        txn_label = "Jama" if t.get("txn_type") == "jama" else "Nikasi"
        results.append({
            "id": t.get("id"), "type": "cash_transaction", "tab": "cashbook",
            "title": f"{t.get('category', '')} ({txn_label})",
            "subtitle": f"Rs.{t.get('amount', 0):,.0f} | {t.get('description', '')[:60]}",
            "date": t.get("date", ""), "data": t
        })

    # 3. Private Paddy
    pvt_cur = db["private_paddy"].find(
        {"$or": [{"party_name": rgx}, {"mandi_name": rgx}]},
        {"_id": 0, "id": 1, "date": 1, "party_name": 1, "mandi_name": 1, "total_amount": 1, "balance": 1}
    ).sort("created_at", -1).limit(limit)
    async for p in pvt_cur:
        results.append({
            "id": p.get("id"), "type": "private_paddy", "tab": "payments",
            "title": f"Private: {p.get('party_name', '')}",
            "subtitle": f"{p.get('mandi_name', '')} | Total: Rs.{p.get('total_amount', 0):,.0f}",
            "date": p.get("date", ""), "data": p
        })

    # 4. Sale Vouchers
    sale_cur = db["sale_vouchers"].find(
        {"$or": [{"party_name": rgx}, {"voucher_no": rgx}, {"truck_no": rgx}]},
        {"_id": 0, "id": 1, "date": 1, "party_name": 1, "voucher_no": 1, "total": 1}
    ).sort("created_at", -1).limit(limit)
    async for s in sale_cur:
        results.append({
            "id": s.get("id"), "type": "sale_voucher", "tab": "vouchers",
            "title": f"Sale: {s.get('party_name', '')}",
            "subtitle": f"Voucher #{s.get('voucher_no', '')} | Rs.{s.get('total', 0):,.0f}",
            "date": s.get("date", ""), "data": s
        })

    # 5. Purchase Vouchers
    pur_cur = db["purchase_vouchers"].find(
        {"$or": [{"party_name": rgx}, {"voucher_no": rgx}, {"truck_no": rgx}]},
        {"_id": 0, "id": 1, "date": 1, "party_name": 1, "voucher_no": 1, "total": 1}
    ).sort("created_at", -1).limit(limit)
    async for pv in pur_cur:
        results.append({
            "id": pv.get("id"), "type": "purchase_voucher", "tab": "vouchers",
            "title": f"Purchase: {pv.get('party_name', '')}",
            "subtitle": f"Voucher #{pv.get('voucher_no', '')} | Rs.{pv.get('total', 0):,.0f}",
            "date": pv.get("date", ""), "data": pv
        })

    # 6. DC Entries
    dc_cur = db["dc_entries"].find(
        {"$or": [{"party_name": rgx}, {"vehicle_no": rgx}, {"lot_no": rgx}]},
        {"_id": 0, "id": 1, "date": 1, "party_name": 1, "vehicle_no": 1, "lot_no": 1, "paddy_type": 1}
    ).sort("created_at", -1).limit(limit)
    async for dc in dc_cur:
        results.append({
            "id": dc.get("id"), "type": "dc_entry", "tab": "dctracker",
            "title": f"DC: {dc.get('party_name', '')}",
            "subtitle": f"Vehicle: {dc.get('vehicle_no', '')} | Lot: {dc.get('lot_no', '')}",
            "date": dc.get("date", ""), "data": dc
        })

    # 7. Staff
    staff_cur = db["staff"].find(
        {"$or": [{"name": rgx}, {"phone": rgx}, {"role": rgx}]},
        {"_id": 0, "id": 1, "name": 1, "phone": 1, "role": 1, "salary": 1}
    ).limit(limit)
    async for st in staff_cur:
        results.append({
            "id": st.get("id"), "type": "staff", "tab": "staff",
            "title": f"Staff: {st.get('name', '')}",
            "subtitle": f"{st.get('role', '')} | Rs.{st.get('salary', 0):,.0f}",
            "date": "", "data": st
        })

    # 8. Milling Entries
    mill_cur = db["milling_entries"].find(
        {"$or": [{"rice_type": rgx}, {"note": rgx}]},
        {"_id": 0, "id": 1, "date": 1, "rice_type": 1, "paddy_input_qntl": 1, "rice_qntl": 1, "outturn_ratio": 1}
    ).sort("created_at", -1).limit(limit)
    async for ml in mill_cur:
        results.append({
            "id": ml.get("id"), "type": "milling", "tab": "milling",
            "title": f"Milling: {ml.get('rice_type', '')}",
            "subtitle": f"Paddy: {ml.get('paddy_input_qntl', 0)}Q | Rice: {ml.get('rice_qntl', 0)}Q | OT: {ml.get('outturn_ratio', 0)}%",
            "date": ml.get("date", ""), "data": ml
        })

    # 9. Diesel Accounts
    diesel_cur = db["diesel_accounts"].find(
        {"$or": [{"truck_no": rgx}, {"pump_name": rgx}, {"description": rgx}]},
        {"_id": 0, "id": 1, "date": 1, "truck_no": 1, "pump_name": 1, "amount": 1}
    ).sort("created_at", -1).limit(limit)
    async for d in diesel_cur:
        results.append({
            "id": d.get("id"), "type": "diesel", "tab": "cashbook",
            "title": f"Diesel: {d.get('truck_no', '')}",
            "subtitle": f"{d.get('pump_name', '')} | Rs.{d.get('amount', 0):,.0f}",
            "date": d.get("date", ""), "data": d
        })

    # 10. Mill Parts Stock
    parts_cur = db["mill_parts_stock"].find(
        {"$or": [{"part_name": rgx}, {"party_name": rgx}]},
        {"_id": 0, "id": 1, "date": 1, "part_name": 1, "party_name": 1, "total_amount": 1, "quantity": 1}
    ).sort("created_at", -1).limit(limit)
    async for mp in parts_cur:
        results.append({
            "id": mp.get("id"), "type": "mill_part", "tab": "mill-parts",
            "title": f"Part: {mp.get('part_name', '')}",
            "subtitle": f"{mp.get('party_name', '')} | Rs.{mp.get('total_amount', 0):,.0f}",
            "date": mp.get("date", ""), "data": mp
        })

    # 11. Hemali Payments
    hemali_cur = db["hemali_payments"].find(
        {"$or": [{"sardar_name": rgx}]},
        {"_id": 0, "id": 1, "date": 1, "sardar_name": 1, "total": 1, "status": 1}
    ).sort("created_at", -1).limit(limit)
    async for h in hemali_cur:
        results.append({
            "id": h.get("id"), "type": "hemali", "tab": "hemali",
            "title": f"Hemali: {h.get('sardar_name', '')}",
            "subtitle": f"Rs.{h.get('total', 0):,.0f} | {h.get('status', '')}",
            "date": h.get("date", ""), "data": h
        })

    # 12. Rice Sales
    rice_cur = db["rice_sales"].find(
        {"$or": [{"buyer_name": rgx}, {"party_name": rgx}]},
        {"_id": 0, "id": 1, "date": 1, "buyer_name": 1, "party_name": 1, "total_amount": 1}
    ).sort("created_at", -1).limit(limit)
    async for rs in rice_cur:
        results.append({
            "id": rs.get("id"), "type": "rice_sale", "tab": "payments",
            "title": f"Rice Sale: {rs.get('buyer_name', '') or rs.get('party_name', '')}",
            "subtitle": f"Rs.{rs.get('total_amount', 0):,.0f}",
            "date": rs.get("date", ""), "data": rs
        })

    # 13. Truck Leases
    lease_cur = db["truck_leases"].find(
        {"$or": [{"truck_no": rgx}, {"owner_name": rgx}]},
        {"_id": 0, "id": 1, "truck_no": 1, "owner_name": 1, "monthly_rent": 1}
    ).limit(limit)
    async for tl in lease_cur:
        results.append({
            "id": tl.get("id"), "type": "truck_lease", "tab": "payments",
            "title": f"Lease: {tl.get('truck_no', '')}",
            "subtitle": f"Owner: {tl.get('owner_name', '')} | Rent: Rs.{tl.get('monthly_rent', 0):,.0f}",
            "date": "", "data": tl
        })

    return {"results": results, "total": len(results), "query": q}
